_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,220 @@
+}
+
++type TestResultSummary struct {
++  ArtifactID  uint      `gorm:"not null;index:testresultsummary_idx_artifact_id" json:"artifactId"`
++  Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  BuildID     uint      `gorm:"not null;index:testresultsummary_idx_build_id" json:"buildId"`
++  Build       *Build    `gorm:"foreignKey:BuildID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  Total       uint      `gorm:"not null" json:"total"`
++  Failed      uint      `gorm:"not null" json:"failed"`
++  Passed      uint      `gorm:"not null" json:"passed"`
++  Skipped     uint      `gorm:"not null" json:"skipped"`
++}
+
++type TestResult struct {
++  ArtifactID  uint              `gorm:"not null;index:testresult_idx_artifact_id" json:"artifactId"`
++  Artifact    *Artifact         `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  Name        string	        `gorm:"not null" json:"name"`
++  Message     string            `gorm:"nullable" json:"message"`
++  StartedOn   *time.Time        `gorm:"nullable;default:NULL;" json:"startedOn" format:"date-time"`
++  CompletedOn *time.Time        `gorm:"nullable;default:NULL;" json:"finishedOn" format:"date-time"`
++  Status      TestResultStatus  `gorm:"not null" enums:"Failed,Passed,Skipped"`
++}
+```
+```go
+// artifact.go
+// new
+type File struct {
+  name string
+  fileName string
+  data []bytes
+}
+// new, /build/{buildid}/test-results
+func (m ArtifactModule) putTestResultsHandler(c *gin.Context) {
+  // parseMultipartFormData to get the files
+  // for each file, parse and store in db
+}
+// new, /build/{buildid}/artifact/{artifactid}/test-results
+func (m ArtifactModule) getBuildArtifactTestResultsHandler(c *gin.Context) {
+  // fetch test results for the specified test
+}
+// new
+func parseMultipartFormData(c *gin.Context) []*File {
+  // ...
+}
+// new
+func parseTRX(file *File) []TestResult, TestResultSummary {
+  // ...
+}
+```
+
+The new endpoint `GET /build/{buildid}/test-results-summary` fetches
+all test result summaries for a build and constructs a json response
+that contains them, and has a summary of them, like:
+
+<details><summary>GET /build/42/test-results-summary</summary>
+
+```json
+{
+  "failed": 4,
+  "passed": 5,
+  "skipped": 0,
+  "artifacts": [
+    {
+      "artifactId": 123,
+      "filename": "MyResults1.trx",
+      "failed": 1,
+      "passed": 3,
+      "skipped": 0
+    },
+    {
+      "artifactId": 124,
+      "filename": "MyResults2.trx",
+      "failed": 3,
+      "passed": 2,
+      "skipped": 0
+    }
+  ]
+}
+```
+</details>
+
+### wharf-web
+
+wharf-web changes to use the new endpoints
+- `GET /build/{buildid}/test-results-summary`\
+  To get summary of all test result summaries for build.
+  
+- `GET /build/{buildid}/artifact/{artifactid}/test-results`\
+  To get test result details for build.
```

Comment:
To not lock ourselves too much into that a test result is an artifact, I suggest instead something like `GET /build/{buildid}/test-results/{testresultid}` _(gosh I dislike that only-lowercase format, but that's an RFC for another day)_

Would also be benefitial with an endpoint with aggregated results from all endpoints on a certain build, as that's most probably how the API will be used anyway as we'll query the test result details for an entire build and not have to query for artifacts first and then have the "n+1" issue again that we have with builds.

Such as: `GET /build/{buildid}/test-result-details`

Reply:
How about
1. `GET /build/{buildid}/test-result-details`
    - Fetches all test result details for all tests for build.
2. `GET /build/{buildid}/test-result-details/{testresultid}` (i.e. testresultsummaryid, artifactid?)
    - Fetches specific all test result details for specific test for build.
3. `GET /build/{buildid}/test-results-summary`
    - Fetches summary of summaries, along with the summaries, for build.
4. `GET /build/{buildid}/test-results-summary/{testresultsummaryid}` (i.e. artifactid)
    - Fetches specific summary for build.

Does that seem right?

`{testresultid}` confused me a little because
- Test results, henceforth test result details, don't have their own unique id currently.
- Test result summaries don't have their own name for their unique id, but since it's a 1-to-1 with Artifact we can use the artifact id.

_(Yeah, and it doesn't get better when it's as long as it is here either 😂)_

Reply:
Sure. Think the 4th endpoint there is a bit frivolous though. You already get that data when querying the 3rd endpoint. And would much rather see the `POST` endpoint return the summary so users don't need to query the one they just uploaded